### PR TITLE
Tests related basic authentication.

### PR DESCRIPTION
**Pull request by lukasz-soszynski-eliatra**
Signed-off-by: Lukasz Soszynski <lukasz.soszynski@eliatra.com>

### Description
[Describe what this change achieves]
* Category (New feature [Tests])
The feature introduces integration tests related to basic authentication. The tests are implemented with the usage of a newly developed test framework.

### Testing
Production code can be treated as a test for tests

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — DISMISSED** [review]
Minor comments, thanks for adding these tests

File: src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java [hunk 1]
```diff
@@ -321,6 +337,29 @@ public String toString() {
...
+		public void assertThatBrowserDoesNotAskUserForCredentials() {
+			Header header = getHeader(HttpHeaders.WWW_AUTHENTICATE);
+			assertThat(header, nullValue());
+		}
+
+		public void assertThatBrowserAskUserForCredentials() {
+			Header header = getHeader(HttpHeaders.WWW_AUTHENTICATE);
+			assertThat(header, notNullValue());
+			assertThat(header.getValue(), containsStringIgnoringCase("basic"));
+		}
```
[the remark this answers is not in the document]
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java]: Done: I moved these methods to test classes.

File: src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java [hunk 2]
```diff
@@ -269,6 +278,13 @@ public Header[] getHeader() {
 			return header;
 		}
 
+		public Header getHeader(String name) {
+			return Arrays.stream(header)
+				.filter(header -> requireNonNull(name, "Header name is mandatory.").equalsIgnoreCase(header.getName()))
+				.findFirst()
+				.orElse(null);
```
[the remark this answers is not in the document]
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java]: done

File: src/integrationTest/java/org/opensearch/security/http/BasicAuthTests.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		.challengingAuthenticator("basic").backend("internal");
+
+	@ClassRule
+	public static final LocalCluster cluster = new LocalCluster.Builder()
+		.clusterManager(ClusterManager.SINGLENODE).anonymousAuth(false)
+		.authc(AUTHC_DOMAIN).users(TEST_USER, SUPER_USER).build();
+
+	@Test
+	public void shouldRespondWith401WhenUserDoesNotExist() {
+		try (TestRestClient client = cluster.getRestClient(NOT_EXISTING_USER, INVALID_PASSWORD)) {
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/security/http/BasicAuthTests.java:58]: Would all of these tests pass if an exception is thrown in `getRestClient`? I see this pattern used in a number of places in the codebase and I also see one instance where we add:

```
} catch (IOException ioe) {
    fail("unexpected exception")
}
```

See: https://github.com/opensearch-project/security/blob/1904db559960fca9d87e334172017c412f3e3d0e/src/test/java/org/opensearch/security/auditlog/compliance/ComplianceAuditlogTest.java#L234-L242

File: src/integrationTest/java/org/opensearch/test/framework/TestSecurityConfig.java [hunk 4]
```diff
@@ -336,8 +337,20 @@ public XContentBuilder toXContent(XContentBuilder xContentBuilder, Params params
 
 	public static class AuthcDomain implements ToXContentObject {
 
+		private static String PUBLIC_KEY = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAoqZbjLUAWc+DZTkinQAdvy1GFjPHPnxheU89hSiWoDD3NOW76H3u3T7cCDdOah2msdxSlBmCBH6wik8qLYkcV8owWukQg3PQmbEhrdPaKo0QCgomWs4nLgtmEYqcZ+QQldd82MdTlQ1QmoQmI9Uxqs1SuaKZASp3Gy19y8su5CV+FZ6BruUw9HELK055sAwl3X7j5ouabXGbcib2goBF3P52LkvbJLuWr5HDZEOeSkwIeqSeMojASM96K5SdotD+HwEyjaTjzRPL2Aa1BEQFWOQ6CFJLyLH7ZStDuPM1mJU1VxIVfMbZrhsUBjAnIhRynmWxML7YlNqkP9j6jyOIYQIDAQAB";
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/TestSecurityConfig.java:340]: Should we be using the new automatic test certificate generation for this?

**Review by cwperks — APPROVED** [review]
This is great @lukasz-soszynski-eliatra! These tests are nice and readable using the new framework. I have a couple of questions, but the changes look good to me.

**Review by DarshitChanpura — COMMENTED** [review]
Ty for this PR @lukasz-soszynski-eliatra !

File: src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (peternied) [reply src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java]: Code doesn't seem updated

File: src/integrationTest/java/org/opensearch/security/http/BasicAuthTests.java (hunk 3 above)
Comment (peternied) [reply src/integrationTest/java/org/opensearch/security/http/BasicAuthTests.java:58]: Read up more here https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

If there is an exception it is bubbled up as if the `getRestClient` was executed inside the `try` block, since there is no `catch(Exception e) {}` block in this test case there is nothing to capture.

TLDR: Exceptions work as expected, the test would fail if there was a problem 🥳

**Review by peternied — APPROVED** [review]

File: src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/cluster/TestRestClient.java]: I think the code was updated: https://github.com/opensearch-project/security/pull/2138/files#diff-091a2aa7163b07f68fa3a40034816a92ba8cf178fd765245cbc802eb79216be9R284

File: src/integrationTest/java/org/opensearch/test/framework/TestSecurityConfig.java (hunk 4 above)
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/TestSecurityConfig.java:340]: Actually, this is not a certificate but just a public key. The public key does not contain any metadata so it never expires.
